### PR TITLE
fix(e2e): Run each test in subscription in different namespace

### PR DESCRIPTION
**Pull request by dinhxuanvu**
All test cases in subscription suite shares the same test namespace.
In some cases, the resource cleanup which doesn't work properly from
previous test causes failure on subsequent test.

This commit will create new namespace for each test so it can be run
in insolation.

Signed-off-by: Vu Dinh <vudinh@outlook.com>

**Description of the change:**

**Motivation for the change:**

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/doc`
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: test/e2e/util_test.go [hunk 1]
```diff
@@ -36,6 +36,7 @@ import (
 	k8scontrollerclient "sigs.k8s.io/controller-runtime/pkg/client"
 
 	gtypes "github.com/onsi/gomega/types"
+	v1 "github.com/operator-framework/api/pkg/operators/v1"
 	"github.com/operator-framework/api/pkg/operators/v1alpha1"
 	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
```
Comment (timflannagan) [inline test/e2e/util_test.go:41]: Feel free to complete ignore this: v1 -> operatorsv1? It also looks like there's duplicate v1alpha1 imports here too.

File: test/e2e/util_test.go [hunk 2]
```diff
@@ -918,3 +919,30 @@ func HaveMessage(goal string) gtypes.GomegaMatcher {
 		return plan.Status.Message
 	}, ContainSubstring(goal))
 }
+
+func SetUpGeneratedTestNamespace(name string) (string, corev1.Namespace) {
+	generatedNamespace := genName(name)
```
Comment (timflannagan) [inline test/e2e/util_test.go:924]: I'd expect this to happen at the call site, but this is fine for now.

File: test/e2e/util_test.go [hunk 3]
```diff
@@ -918,3 +919,30 @@ func HaveMessage(goal string) gtypes.GomegaMatcher {
...
+		},
+		Spec: v1.OperatorGroupSpec{
+			TargetNamespaces: []string{ns.GetName()},
+		},
+	}
+	Eventually(func() error {
+		return ctx.Ctx().Client().Create(context.Background(), &og)
+	}).Should(Succeed())
+
+	return generatedNamespace, ns
```
Comment (timflannagan) [inline test/e2e/util_test.go:947]: Hmm, not opposed to this function signature, but you can access the generated namespace's name through `ns.GetName()` type operations (or indexing into the metadata.Name field directly).

File: test/e2e/util_test.go [hunk 4]
```diff
@@ -918,3 +919,30 @@ func HaveMessage(goal string) gtypes.GomegaMatcher {
...
+	og := v1.OperatorGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-operatorgroup", ns.GetName()),
+			Namespace: ns.GetName(),
+		},
+		Spec: v1.OperatorGroupSpec{
+			TargetNamespaces: []string{ns.GetName()},
+		},
+	}
+	Eventually(func() error {
```
Comment (timflannagan) [inline test/e2e/util_test.go:943]: :+1:

File: test/e2e/util_test.go [hunk 5]
```diff
@@ -918,3 +919,30 @@ func HaveMessage(goal string) gtypes.GomegaMatcher {
 		return plan.Status.Message
 	}, ContainSubstring(goal))
 }
+
+func SetUpGeneratedTestNamespace(name string) (string, corev1.Namespace) {
+	generatedNamespace := genName(name)
+	ns := corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
```
Comment (timflannagan) [inline test/e2e/util_test.go:926]: Mainly curious on when this is necessary: do we need to specify TypeMeta here as well?

**Review by timflannagan — COMMENTED** [review]
Thanks, this should be help reduce many of the existing subscription-related e2e failures that are due to test pollution affect unrelated tests, and making debugging those failures less difficult. I only had one real comment: any reason we need to supply both a generatedNamespace and a corev1.Namespace as output variables from the SetUpGeneratedTestNamespace function? We should be able to infer the generated namespace by indexing into the corev1.Namespace's fields.

**Comment by dinhxuanvu** [comment]
The commit is included in #2438 which is merged.
/close

**Pull request closed without merging**